hypothesis-python/src/hypothesis/extra/github_actions.py

```diff
@@ -0,0 +1,156 @@
+        self.token = getenv("GITHUB_TOKEN")
+
+        # We want to be lazy per conftest initialization
+        self._artifact_downloaded = False
+
+    def __repr__(self) -> str:
+        return f"GitHubArtifactDatabase(owner={self.owner}, repo={self.repo}, artifact_name={self.artifact_name})"
+
+    def _fetch_artifact(self) -> None:
+        if self._artifact_downloaded:
+            return
+
+        # Get the latest artifact from the GitHub API
+        try:
+            res = requests.get(
+                f"https://api.github.com/repos/{self.owner}/{self.repo}/actions/artifacts",
+                headers={
+                    "Accept": "application/vnd.github+json",
+                    "X-GitHub-Api-Version": "2022-11-28 ",
+                    "Authorization": f"Bearer {self.token}",
+                },
+            )
+            res.raise_for_status()
+            artifacts: list[dict] = res.json()["artifacts"]
+        except requests.exceptions.ConnectionError as e:
+            raise RuntimeError(
+                "Could not connect to GitHub to get the latest artifact."
+            ) from e
+        except requests.exceptions.HTTPError as e:
+            # TODO: Be more granular
+            raise RuntimeError(
+                "Could not get the latest artifact from GitHub. "
+                "Check that the repository exists and that you've provided a valid token (GITHUB_TOKEN)."
+            ) from e
+
+        # Get the latest artifact from the list
+        artifact = sorted(
+            filter(lambda a: a["name"] == self.artifact_name, artifacts),
+            key=lambda a: a["created_at"],
+        )[-1]
+
+        # Download and extract the artifact into .hypothesis/ci
+        with tempfile.NamedTemporaryFile() as f:
+            try:
+                req = requests.get(
+                    artifact["archive_download_url"],
+                    headers={
+                        "Accept": "application/vnd.github+json",
+                        "X-GitHub-Api-Version": "2022-11-28",
+                        "Authorization": f"Bearer {self.token}",
+                    },
+                    stream=True,
+                    allow_redirects=True,
+                )
+                req.raise_for_status()
+            except requests.exceptions.ConnectionError as e:
+                raise RuntimeError(
+                    "Could not connect to GitHub to download the artifact."
+                ) from e
+            except requests.exceptions.HTTPError as e:
+                raise RuntimeError(
+                    "Could not get the latest artifact from GitHub. "
+                    "Check that the repository exists and that you've provided a valid token (GITHUB_TOKEN)."
+                ) from e
+
+            f.write(req.content)
+
+            # Extract the artifact
+            mkdir_p(self.path)
+            shutil.unpack_archive(f.name, self.path, "zip")
```

Comment:
I'd prefer to do this lazily in-memory using https://docs.python.org/3/library/zipfile.html - it should be faster, as well as requiring less storage space.

(downside is that we're basically inventing a `ZipfileDatabase`, but that does make it easier to test everything-but-downloading)

Reply:
Oh, I'll have to figure out the low-level database primitives then.